### PR TITLE
AB#363 Generate certificates as secret

### PR DESCRIPTION
**Pull request by Nirusu**
This PR adds support to define a certificate as secret which should be generated.

This can be done by following the x509.Certificate golang object notation. Every field from the object can be defined in the manifest, the generator will take most values and overwrite some of them and generate a key pair + certificate based on them.

Right now, this is not completely done yet as tests are failing, documentation still needs to be provided etc., but it should be ready for review.

### CODE REVIEW

**Comment by Nirusu** [comment]
Tests are now passing with the latest change to get rid of the pointer in the struct. Might be worth to check if there are still some good test cases I could add, this feature feels really fragile to work on.

File: coordinator/core/core.go
Comment (thomasten) [inline coordinator/core/core.go:496]:
```suggestion
func (c *Core) generateCertificateForSecret(secret Secret, key crypto.PublicKey) (x509.Certificate, error) {
```

File: coordinator/core/core.go
Comment (thomasten) [inline coordinator/core/core.go:498]:
```suggestion
	var template x509.Certificate
```

File: coordinator/core/core.go
Comment (thomasten) [inline coordinator/core/core.go:508]:
```suggestion
		template.IPAddresses = util.DefaultCertificateIPAddresses
```
and add `var DefaultCertificateIPAddresses = []net.IP{net.IPv4(127, 0, 0, 1), net.IPv6loopback}` to util/tls.go and also replace all other occurrences

File: coordinator/core/core.go
Comment (thomasten) [inline coordinator/core/core.go:505]:
```suggestion
		template.DNSNames = []string{"localhost"}
```

File: coordinator/core/core.go
Comment (thomasten) [inline coordinator/core/core.go:520]:
```suggestion
		template.SerialNumber, err = util.GenerateCertificateSerialNumber()
```

File: coordinator/core/core.go
Comment (thomasten) [inline coordinator/core/core.go:537]:
```suggestion
	if secret.ValidFor == 0 {
		secret.ValidFor = 365
	}
	template.NotAfter = time.Now().Add(time.Hour * 24 * time.Duration(secret.ValidFor))
```

File: coordinator/core/core_test.go
Comment (thomasten) [inline coordinator/core/core_test.go:178]: These are always true. Instead, check some fields of the certificates.

File: coordinator/core/marbleapi_test.go
Comment (thomasten) [inline coordinator/core/marbleapi_test.go:207]:
```suggestion
```

File: coordinator/core/marbleapi_test.go
Comment (thomasten) [inline coordinator/core/marbleapi_test.go:211]: Check that NotAfter-NotBefore is 7 days

File: coordinator/core/marbleapi_test.go
Comment (thomasten) [inline coordinator/core/marbleapi_test.go:227]:
```suggestion
	key, err := rsa.GenerateKey(rand.Reader, 2048)
```

File: coordinator/core/marbleapi_test.go
Comment (thomasten) [inline coordinator/core/marbleapi_test.go:230]:
```suggestion
	require.NoError(err)
```

File: coordinator/core/marbleapi_test.go
Comment (thomasten) [inline coordinator/core/marbleapi_test.go:298]:
```suggestion
	require.NotNil(p)
```

**Review by thomasten — CHANGES_REQUESTED** [review]

File: coordinator/core/core.go
Comment (thomasten) [inline coordinator/core/core.go:502]:
```suggestion
	// Generate certificate with given public key
	secretCertRaw, err := x509.CreateCertificate(rand.Reader, &template, c.cert, pubKey, c.privk)
	if err != nil {
```

File: coordinator/core/marbleapi_test.go
Comment (thomasten) [inline coordinator/core/marbleapi_test.go:290]: also add a test for `{{ raw .Secrets.testcertificate.Cert }}` (and implement the feature)

**Review by thomasten — CHANGES_REQUESTED** [review]

File: test/manifests.go
Comment (Nirusu) [reply test/manifests.go:115]: Not really sure about this one. This creates ambiguity with the x509.Certificate fields (ValidFor vs. NotAfter), plus directly addressing or setting the certificate as an object becomes uglier.

File: coordinator/core/manifest.go
Comment (thomasten) [inline coordinator/core/manifest.go:129]: we should also use a switch here

File: coordinator/core/core.go
Comment (thomasten) [inline coordinator/core/core.go:483]:
```suggestion
	} else if secret.ValidFor != 0 {
```

**Review by thomasten — CHANGES_REQUESTED** [review]

**Review by m1ghtym0 — APPROVED** [review]
LGTM:-)
We should make a new release with this one.
I've seen the docs already get updated👍

**Comment by Nirusu** [comment]
I'd suggest we wait for a new release until the "unique"/"marble only" secrets are merged, then this feels more complete, I guess.

I'll open the PR for it right when this one is merged into master, maybe we can get everything done related to secrets this week :)

**Comment by m1ghtym0** [comment]
> I'd suggest we wait for a new release until the "unique"/"marble only" secrets are merged, then this feels more complete, I guess.
> 
> I'll open the PR for it right when this one is merged into master, maybe we can get everything done related to secrets this week :)

Oh, wasn't aware there is more open work in secrets. Sure let's wait until secrets is complete.

**Review by thomasten — APPROVED** [review]

**Pull request merged**
